src/jit/morph.cpp

```diff
@@ -12453,22 +12443,12 @@ GenTree* Compiler::fgMorphSmpOpOptional(GenTreeOp* tree)
                     }
 
                     /* Special case: "x |= -1" and "x &= 0" */
-                    if  (cmop == GT_AND || cmop == GT_OR)
+                    if (((cmop == GT_AND) && op2->gtOp.gtOp2->IsIntegralConst(0)) ||
+                        ((cmop == GT_OR)  && op2->gtOp.gtOp2->IsIntegralConst(-1)))
                     {
-                        if  (op2->gtOp.gtOp2->IsCnsIntOrI())
-                        {
-                            ssize_t     icon = op2->gtOp.gtOp2->gtIntCon.gtIconVal;
-
-                            noway_assert(typ <= TYP_UINT);
```

Comment:
This noway_assert got deleted. Looks like this optimization was done only for TYP_INT case on 32-bit but not for long/ulong.  With this change this optimization gets enabled for TYP_LONG/ULONG. Good thing though.  May be we need to modify this noway_assert to 

noway_assert(typ <= TYP_ULONG);